modules/objdetect/src/aruco/aruco_detector.cpp

```diff
@@ -651,9 +657,257 @@ struct ArucoDetector::ArucoDetectorImpl {
+                            findCornerInPyrImage(scale_init, closest_pyr_image_idx, grey_pyramid, Mat(candidates[i]), detectorParams);
+                        } else {
+                            int cornerRefinementWinSize = std::max(1, cvRound(detectorParams.relativeCornerRefinmentWinSize*
+                                        getAverageModuleSize(candidates[i], dictionary.markerSize, detectorParams.markerBorderBits)));
+                            cornerRefinementWinSize = min(cornerRefinementWinSize, detectorParams.cornerRefinementWinSize);
+                            cornerSubPix(grey, Mat(candidates[i]), Size(cornerRefinementWinSize, cornerRefinementWinSize), Size(-1, -1),
+                                    TermCriteria(TermCriteria::MAX_ITER | TermCriteria::EPS,
+                                        detectorParams.cornerRefinementMaxIterations,
+                                        detectorParams.cornerRefinementMinAccuracy));
+                        }
+                    }
+                });
+            }
+        } else if (DictionaryMode::Multi == dictMode) {
+            unordered_set<int> uniqueMarkerSizes;
+            for (const Dictionary& dictionary : dictionaries) {
+                uniqueMarkerSizes.insert(dictionary.markerSize);
+            }
+
+            // create at max 4 marker candidate trees for each dictionary size
+            vector<vector<MarkerCandidateTree>> candidatesPerDictionarySize = {{}, {}, {}, {}};
+            for (int markerSize : uniqueMarkerSizes) {
+                // min marker size is 4, so subtract 4 to get index
+                const auto dictionarySizeIndex = markerSize - 4;
+                // copy candidates
+                vector<vector<Point2f>> candidatesCopy = candidates;
+                vector<vector<Point> > contoursCopy = contours;
+                candidatesPerDictionarySize[dictionarySizeIndex] = filterTooCloseCandidates(candidatesCopy, contoursCopy, markerSize);
+            }
+            candidates.clear();
+            contours.clear();
+
+            /// STEP 2: Check candidate codification (identify markers)
+            int dictIndex = 0;
+            for (const Dictionary&  currentDictionary : dictionaries) {
+                const auto dictionarySizeIndex = currentDictionary.markerSize - 4;
+                // temporary variable to store the current candidates
+                vector<vector<Point2f>> currentCandidates;
+                identifyCandidates(grey, grey_pyramid, candidatesPerDictionarySize[dictionarySizeIndex], currentCandidates, contours,
+                        ids, currentDictionary, rejectedImgPoints);
+                if (_dictIndices.needed()) {
+                    dictIndices.insert(dictIndices.end(), currentCandidates.size(), dictIndex);
+                }
+
+                /// STEP 3: Corner refinement :: use corner subpix
+                if (detectorParams.cornerRefinementMethod == (int)CORNER_REFINE_SUBPIX) {
```

Comment:
Is it possible to extract this large block to separate function to reduce code duplication?